packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -445,6 +454,273 @@ class CupertinoPopupSurface extends StatelessWidget {
+  /// the drag starts, which is useful to enhance responsiveness.
+  ValueSetter<Offset>? onResponsiveUpdate;
+
+  /// Called whenever the primary pointer is lifted regardless of whether drag
+  /// has started.
+  ///
+  /// The parameter is the global position of the primary pointer.
+  ///
+  /// This allows the caller to track the primary pointer's location before
+  /// the drag starts, which is useful to enhance responsiveness.
+  ValueSetter<Offset>? onResponsiveEnd;
+
+  int? _primaryPointer;
+
+  @override
+  void addAllowedPointer(PointerDownEvent event) {
+    _primaryPointer ??= event.pointer;
+    super.addAllowedPointer(event);
+  }
+
+  @override
+  void rejectGesture(int pointer) {
+    if (pointer == _primaryPointer) {
+      _primaryPointer = null;
+    }
+    super.rejectGesture(pointer);
+  }
+
+  @override
+  void handleEvent(PointerEvent event) {
+    if (event.pointer == _primaryPointer) {
+      if (event is PointerMoveEvent) {
+        onResponsiveUpdate?.call(event.position);
+      }
+      // If this gesture has a competing gesture (such as scrolling), and the
+      // pointer has not moved far enough to get this panning accepted, a
+      // pointer up event should still be considered as an accepted tap up.
+      // Manually accept this gesture here, which triggers onDragEnd.
+      if (event is PointerUpEvent) {
+        resolve(GestureDisposition.accepted);
+        stopTrackingPointer(_primaryPointer!);
+        onResponsiveEnd?.call(event.position);
+      } else {
+        super.handleEvent(event);
+      }
+      if (event is PointerUpEvent || event is PointerCancelEvent) {
+        _primaryPointer = null;
+      }
+    }
+  }
+
+  @override
+  String get debugDescription => 'tap slide';
+}
+
+// A region (typically a button) that can receive entering, exiting, and
+// updating events of a "sliding tap" gesture.
+//
+// Some Cupertino widgets, such as action sheets or dialogs, allow the user to
+// select buttons using "sliding taps", where the user can drag around after
+// pressing on the screen, and whichever button the drag ends in is selected.
+//
+// This class is used to define the regions that sliding taps recognize. This
+// class must be provided to a `MetaData` widget as `data`, and is typically
+// implemented by a widget state class. When an eligible dragging gesture
+// enters, leaves, or ends this `MetaData` widget, corresponding methods of this
+// class will be called.
+//
+// Multiple `_ActionSheetDragAvatar`s might be nested.
+// `_AvatarSelectionGestureRecognizer` uses a simple algorithm that only
+// compares if the inner-most avatar has changed (which suffices our use case).
+// Semantically, this means that all outer avatars will be treated as identical
+// to the inner-most one, i.e. when the pointer enters or leaves an avatar, the
+// corresponding method will be called on all avatars that nest it.
+abstract class _ActionSheetDragAvatar {
```

Comment:
Yeah, I've renamed it to `_ActionSheetSlideTarget`. (At the beginning of development I was looking at `DragAvatar` but the code has deviated a lot since then.)